domain/application/state/application_endpoint.go

```diff
@@ -94,38 +133,33 @@ func (st *State) insertApplicationEndpoints(ctx context.Context, tx *sqlair.TX,
 		return errors.Errorf("checking charm relation: %w", err)
 	}
 
-	// Update default space.
-	if err := st.updateDefaultSpace(ctx, tx, params.appID, params.bindings); err != nil {
-		return errors.Errorf("updating default space: %w", err)
-	}
-
-	// Insert endpoints.
+	// Upsert endpoints.
 	for _, relation := range relations {
-		if err := st.insertApplicationEndpoint(ctx, tx, params.appID, relation, params.bindings); err != nil {
-			return errors.Errorf("inserting application endpoint: %w", err)
+		if err := st.upsertApplicationEndpoint(ctx, tx, params.appID, relation, params.bindings); err != nil {
+			return errors.Errorf("upserting application endpoint: %w", err)
 		}
 	}
 
-	// Insert extra binding
+	// Upsert extra binding
 	for _, binding := range extrabindings {
-		if err := st.insertApplicationExtraBinding(ctx, tx, params.appID, binding, params.bindings); err != nil {
-			return errors.Errorf("inserting application endpoint: %w", err)
+		if err := st.upsertApplicationExtraBinding(ctx, tx, params.appID, binding, params.bindings); err != nil {
+			return errors.Errorf("upserting application endpoint: %w", err)
 		}
 	}
 
 	return nil
 }
 
-// insertApplicationEndpoint inserts an application endpoint into the database,
+// upsertApplicationEndpoint upserts an application endpoint into the database,
 // associating it with a relation and space.
-func (st *State) insertApplicationEndpoint(
+func (st *State) upsertApplicationEndpoint(
```

Comment:
It is not correct to recruit this as an upsert for merging bindings. Bindings are user input; the endpoints come from the charm definition.

The action of merging bindings should only do updates.

Leave this as it was, validate the incoming bindings map against the existing endpoints, and set only the bound spaces.